File: rd-core/src/commonMain/kotlin/com/jetbrains/rider/util/threading/SynchronousScheduler.kt
package com.jetbrains.rider.util.threading

import com.jetbrains.rider.util.reactive.IScheduler
import com.jetbrains.rider.util.threadLocalWithInitial


object SynchronousScheduler : IScheduler {
    override fun flush() {

    }

    override fun queue(action: () -> Unit) {
        active.set(active.get() + 1)
        try {
            action()
        } finally {
            active.set(active.get() - 1)
        }
    }

    private val active =  threadLocalWithInitial { 0 }

    override val isActive: Boolean
        get() = active.get() > 0



}

File: rd-cpp/src/rd_framework_cpp/src/main/scheduler/SynchronousScheduler.cpp
#include "SynchronousScheduler.h"

#include "guards.h"

namespace rd
{
static thread_local int32_t SynchronousScheduler_active_count = 0;

void SynchronousScheduler::queue(std::function<void()> action)
{
	util::increment_guard<int32_t> guard(SynchronousScheduler_active_count);
	action();
}

void SynchronousScheduler::flush()
{
}

bool SynchronousScheduler::is_active() const
{
	return SynchronousScheduler_active_count > 0;
}
}	 // namespace rd